### PR TITLE
Adding API call getdeploymentflags for soft fork deployments (BIP9).

### PR DESCRIPTION
**Pull request by thefazzer**
New API call: `http://localhost:37220/api/Consensus/getdeploymentflags`

- Get the threshold states of softforks currently being deployed.
- Allowable states: Defined, Started, LockedIn, Failed, Active.
- Returns Json formatted type with Deployment Index, State Value, Human readable threshold state.

Example:
`"deploymentIndex": 0,
"stateValue": 3,
"thresholdState": "Failed"
`

Preferred: @dangershony @zeptin

### CODE REVIEW

**Comment by dangershony** [comment]
I wonder if we could also get the number of blocks that have signaled a deployment

**Comment by zeptin** [comment]
It's possible, but it looks like the return type of `ThresholdConditionCache.GetState()` would need to be enriched to include additional data from the state computations. It isn't directly called in too many spots so that seems feasible.

**Comment by dangershony** [comment]
Ok so we should probably do it, and as part of this PR, so lets keep it open til one of us can pick it up.

**Comment by thefazzer** [comment]
@dangershony What about the following?  This enriches the state information with the number of blocks in each threshold state per deployment:

`[{"deploymentIndex":0,"blocksDefined":0,"blocksStarted":0,"blocksLockedIn":0,"blocksFailed":107,"blocksActive":0,"stateValue":3,"thresholdState":"Failed"},{"deploymentIndex":1,"blocksDefined":1,"blocksStarted":0,"blocksLockedIn":0,"blocksFailed":0,"blocksActive":0,"stateValue":0,"thresholdState":"Defined"},{"deploymentIndex":2,"blocksDefined":1,"blocksStarted":0,"blocksLockedIn":0,"blocksFailed":0,"blocksActive":0,"stateValue":0,"thresholdState":"Defined"}]`

@zeptin As you point out we have a ton of voids (caused by null entries).  Will look into why.

**Comment by thefazzer** [comment]
`{
    "deploymentIndex": 1,
    "blocksDefined": 358,
    "blocksStarted": 22,
    "blocksLockedIn": 1,
    "blocksFailed": 0,
    "blocksActive": 1,
    "stateValue": 4,
    "thresholdState": "Active"
  },`

@zeptin FYI the nulls mentioned were due to chain not being downloaded.  The above returned against testnet after activation for start date for deployment index 1.

File: src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs [hunk 1]
```diff
@@ -45,6 +49,36 @@ public uint256 GetBestBlockHashRPC()
...
+        public IActionResult GetDeploymentFlags()
+        {
+            try
+            {
+               var rule = this.ConsensusManager.ConsensusRules.GetRule<SetActivationDeploymentsFullValidationRule>();
+               
+               // Ensure threshold conditions cached.
+               ThresholdState[] thresholdStates = rule.Parent.NodeDeployments.BIP9.GetStates(this.ChainState.ConsensusTip.Previous);
+
+               object metrics = rule.Parent.NodeDeployments.BIP9.EnrichStatesWithBlockMetrics(thresholdStates);
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs]: Lets use defined models for consistency?

File: src/Stratis.Bitcoin/Base/Deployments/ThresholdConditionCache.cs [hunk 2]
```diff
@@ -172,9 +177,97 @@ public ThresholdState GetState(ChainedHeader indexPrev, int deployment)
                 }
 
                 this.Set(indexPrev?.HashBlock, deployment, state = stateNext);
+
+                // FAZZ
```
Comment (zeptin) [inline src/Stratis.Bitcoin/Base/Deployments/ThresholdConditionCache.cs]: Ahem

File: src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs [hunk 3]
```diff
@@ -45,6 +49,36 @@ public uint256 GetBestBlockHashRPC()
...
 
+        /// <summary>
+        /// Get the threshold states of softforks currently being deployed.
+        /// Allowable states are: Defined, Started, LockedIn, Failed, Active.
+        /// </summary>
+        /// <returns>Json formatted type with Deployment Index <see cref="int"/>, 
+        /// State Value <see cref="ThresholdState"/>, human readable Threshold State <see cref="string"/>
+        /// Returns <see cref="IActionResult"/> formatted error if fails.
+        /// </returns>.
+        [Route("api/[controller]/getdeploymentflags")]
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs]: you can remove the `get` from the endpoint as the verb used is GET.

**Comment by thefazzer** [comment]
@dangershony @zeptin Latest output from testnet.

Notes: threshold is present to compare to votes, also timeStart/medianTimePast/timeTimeOut

`[
  {
    "deploymentIndex": 2,
    "blocksDefined": 0,
    "blocksStarted": 45,
    "blocksLockedIn": 0,
    "blocksFailed": 0,
    "blocksActive": 0,
    "medianTimePast": 1547034400,
    "timeStart": 1541030400,
    "timeTimeOut": 1548979200,
    "threshold": 1916,
    "votes": 1897,
    "stateValue": 1,
    "thresholdState": "Started"
  }
]`

**Comment by thefazzer** [comment]
@zeptin Fixed that issue you had on Friday at startup

File: src/Stratis.Bitcoin/Base/Deployments/ThresholdConditionCache.cs (hunk 2 above)
Comment (dangershony) [reply src/Stratis.Bitcoin/Base/Deployments/ThresholdConditionCache.cs]: Marking his territory 😄

**Comment by thefazzer** [comment]
@dangershony Testnet output - votes refer to those within confirmation window.

`[
  {
    "deploymentIndex": 2,
    "stateValue": 1,
    "thresholdState": "Started",
    "height": 728723,
    "periodStartsHeight": 727775,
    "periodEndsHeight": 729791,
    "votes": 27,
    "threshold": 1916,
    "timeStart": 1541030400,
    "timeTimeOut": 1548979200
  }
]`

**Comment by thefazzer** [comment]
@bokobza Could you please take a look - any outstanding issues with the PR or comments?

**Comment by zeptin** [comment]
@bokobza It looks like the data coming out is well-formed:

```
[
  {
    "deploymentIndex": 2,
    "stateValue": 1,
    "thresholdState": "Started",
    "height": 738404,
    "periodStartsHeight": 737855,
    "periodEndsHeight": 739871,
    "votes": 194,
    "threshold": 1916,
    "timeStart": "2018-11-01T00:00:00",
    "timeTimeOut": "2019-02-01T00:00:00"
  }
]
```

I am attempting to independently verify that the vote count is accurate, but this looks good.

File: src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs [hunk 4]
```diff
@@ -45,6 +49,36 @@ public uint256 GetBestBlockHashRPC()
...
+        public IActionResult DeploymentFlags()
+        {
+            try
+            {
+               var rule = this.ConsensusManager.ConsensusRules.GetRule<SetActivationDeploymentsFullValidationRule>();
+
+                // Ensure threshold conditions cached.
+                ThresholdState[] thresholdStates = rule.Parent.NodeDeployments.BIP9.GetStates(this.ChainState.ConsensusTip.Previous);
+
+                object metrics = rule.Parent.NodeDeployments.BIP9.GetThresholdStateMetrics(this.ChainState.ConsensusTip.Previous, thresholdStates);
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs]: why `object`?

File: src/Stratis.Bitcoin/Base/Deployments/ThresholdConditionCache.cs [hunk 5]
```diff
@@ -58,6 +59,49 @@ public ThresholdState[] GetStates(ChainedHeader pindexPrev)
             return array;
         }
 
+        public object GetThresholdStateMetrics(ChainedHeader indexPrev, ThresholdState[] thresholdStates)
```
Comment (bokobza) [inline src/Stratis.Bitcoin/Base/Deployments/ThresholdConditionCache.cs]: Why `object`?

File: src/Stratis.Bitcoin/Base/Deployments/ThresholdConditionCache.cs [hunk 6]
```diff
@@ -58,6 +59,49 @@ public ThresholdState[] GetStates(ChainedHeader pindexPrev)
...
+        {
+            var nonNullDeployments = new List<int>();
+            for (int i = 0; i < this.consensus.BIP9Deployments.Length; i++)
```
Comment (bokobza) [inline src/Stratis.Bitcoin/Base/Deployments/ThresholdConditionCache.cs]: It feels a bit wasteful to keep the deployment in an array and then in a new loop, loop around all the deployments and skip over the null ones.
Any chance we can have one loop?

File: src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs [hunk 7]
```diff
@@ -0,0 +1,46 @@
+﻿using System;
+
+namespace Stratis.Bitcoin.Base.Deployments.Models
+{
+    /// <summary>
+    /// Class representing the activation states with the count of blocks in each state.
+    /// </summary>
+    public class ThresholdStateModel
+    {
+        public int DeploymentIndex { get; }
```
Comment (bokobza) [inline src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs]: Does the index have any meaning?

File: src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+        public int Votes { get; set; }
+
+        public int Threshold { get; }
+
+        public DateTime? TimeStart { get; }
+
+        public DateTime? TimeTimeOut { get; }
+
+        public ThresholdStateModel(int deploymentIndex, int votes, DateTime? timeStart, DateTime? timeTimeOut, int threshold, int height, int periodStartsHeight, int periodEndsHeight, ThresholdState stateValue, string thresholdState)
```
Comment (bokobza) [inline src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs]: You don't really need a ctor and a default ctor here, you can just have `get; set;` on the property and construct the object when you need it, like we do in the other model classes.

File: src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs [hunk 9]
```diff
@@ -45,6 +49,36 @@ public uint256 GetBestBlockHashRPC()
...
+        /// State Value <see cref="ThresholdState"/>, human readable Threshold State <see cref="string"/>
+        /// Returns <see cref="IActionResult"/> formatted error if fails.
+        /// </returns>.
+        [Route("api/[controller]/deploymentflags")]
+        [HttpGet]
+        public IActionResult DeploymentFlags()
+        {
+            try
+            {
+               var rule = this.ConsensusManager.ConsensusRules.GetRule<SetActivationDeploymentsFullValidationRule>();
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs]: Why getting a rule if we need rule engine?

File: src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs [hunk 10]
```diff
@@ -45,6 +49,36 @@ public uint256 GetBestBlockHashRPC()
...
+                // Ensure threshold conditions cached.
+                ThresholdState[] thresholdStates = rule.Parent.NodeDeployments.BIP9.GetStates(this.ChainState.ConsensusTip.Previous);
+
+                object metrics = rule.Parent.NodeDeployments.BIP9.GetThresholdStateMetrics(this.ChainState.ConsensusTip.Previous, thresholdStates);
+
+                return this.Json(metrics);
+            }
+            catch (Exception e)
+            {
+                this.logger.LogError("Exception occurred: {0}", e.ToString());
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs]: exit log missing

File: src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+        public ThresholdState? StateValue { get; set; }
+
+        public string ThresholdState { get; set; }
+
+        public int Height { get; }
+
+        public int PeriodStartsHeight { get; }
+
+        public int PeriodEndsHeight { get; }
```
Comment (noescape00) [inline src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs]: Please comment all the properties

File: src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs (hunk 9 above)
Comment (thefazzer) [reply src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs]: Updated to rule engine.

File: src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs (hunk 4 above)
Comment (thefazzer) [reply src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs]: Updated object to actual type.

File: src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs (hunk 10 above)
Comment (thefazzer) [reply src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs]: Added.

File: src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs (hunk 7 above)
Comment (thefazzer) [reply src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs]: Deployment index for this soft fork.

File: src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs (hunk 11 above)
Comment (thefazzer) [reply src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs]: Done.

File: src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs (hunk 8 above)
Comment (thefazzer) [reply src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs]: Replaced with mutable object.

File: src/Stratis.Bitcoin/Base/Deployments/ThresholdConditionCache.cs (hunk 5 above)
Comment (thefazzer) [reply src/Stratis.Bitcoin/Base/Deployments/ThresholdConditionCache.cs]: As above.

File: src/Stratis.Bitcoin/Base/Deployments/ThresholdConditionCache.cs (hunk 6 above)
Comment (thefazzer) [reply src/Stratis.Bitcoin/Base/Deployments/ThresholdConditionCache.cs]: Logic refined as suggested.

File: src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs [hunk 12]
```diff
@@ -1,11 +1,17 @@
 ﻿using System;
+using System.Collections.Generic;
+using System.Linq;
```
Comment (quantumagi) [inline src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs]: Not used

File: src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs [hunk 13]
```diff
@@ -1,11 +1,17 @@
...
 using System.Net;
 using Microsoft.AspNetCore.Mvc;
 using Microsoft.Extensions.Logging;
 using NBitcoin;
 using Stratis.Bitcoin.Base;
+using Stratis.Bitcoin.Base.Deployments;
+using Stratis.Bitcoin.Base.Deployments.Models;
 using Stratis.Bitcoin.Consensus;
+using Stratis.Bitcoin.Consensus.Rules;
```
Comment (quantumagi) [inline src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs]: Not used

File: src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs [hunk 14]
```diff
@@ -45,6 +51,36 @@ public uint256 GetBestBlockHashRPC()
             return this.ChainState.ConsensusTip?.HashBlock;
         }
 
+        /// <summary>
+        /// Get the threshold states of softforks currently being deployed.
+        /// Allowable states are: Defined, Started, LockedIn, Failed, Active.
+        /// </summary>
+        /// <returns>Json formatted type with Deployment Index <see cref="int"/>,
```
Comment (quantumagi) [inline src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs]: Redundant whitespace at the end of this line.

File: src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs [hunk 15]
```diff
@@ -45,6 +51,36 @@ public uint256 GetBestBlockHashRPC()
...
+
+                return this.Json(metrics);
+            }
+            catch (Exception e)
+            {
+                this.logger.LogError("Exception occurred: {0}", e.ToString());
+                return ErrorHelpers.BuildErrorResponse(HttpStatusCode.BadRequest, e.Message, e.ToString());
+            }
+        }
+
```
Comment (quantumagi) [inline src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs]: Redundant whitespace in this line.

File: src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs [hunk 16]
```diff
@@ -45,6 +51,36 @@ public uint256 GetBestBlockHashRPC()
...
+        /// State Value <see cref="ThresholdState"/>, human readable Threshold State <see cref="string"/>
+        /// Returns <see cref="IActionResult"/> formatted error if fails.
+        /// </returns>.
+        [Route("api/[controller]/deploymentflags")]
+        [HttpGet]
+        public IActionResult DeploymentFlags()
+        {
+            try
+            {
+               ConsensusRuleEngine ruleEngine = this.ConsensusManager.ConsensusRules.GetRule<SetActivationDeploymentsFullValidationRule>().Parent;
```
Comment (quantumagi) [inline src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs]: Incorrect indentation.

File: src/Stratis.Bitcoin/Base/Deployments/ThresholdConditionCache.cs [hunk 17]
```diff
@@ -58,6 +59,55 @@ public ThresholdState[] GetStates(ChainedHeader pindexPrev)
             return array;
         }
 
+        public List<ThresholdStateModel> GetThresholdStateMetrics(ChainedHeader indexPrev, ThresholdState[] thresholdStates)
```
Comment (quantumagi) [inline src/Stratis.Bitcoin/Base/Deployments/ThresholdConditionCache.cs]: Missing XML documentation for this method.

File: src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs [hunk 18]
```diff
@@ -45,6 +51,36 @@ public uint256 GetBestBlockHashRPC()
             return this.ChainState.ConsensusTip?.HashBlock;
         }
 
+        /// <summary>
+        /// Get the threshold states of softforks currently being deployed.
+        /// Allowable states are: Defined, Started, LockedIn, Failed, Active.
+        /// </summary>
+        /// <returns>Json formatted type with Deployment Index <see cref="int"/>, 
+        /// State Value <see cref="ThresholdState"/>, human readable Threshold State <see cref="string"/>
```
Comment (quantumagi) [inline src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs]: Missing period at the end of this line.

File: src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs [hunk 19]
```diff
@@ -0,0 +1,71 @@
+﻿using System;
+using Newtonsoft.Json;
+
+namespace Stratis.Bitcoin.Base.Deployments.Models
+{
+    /// <summary>
+    /// Class representing the activation states with the count of blocks in each state.
```
Comment (quantumagi) [inline src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs]: `/// Class representing the current activation state of a deployment.`

File: src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+        /// <summary>
+        /// Height of the the block with this threshold state.
+        /// </summary>
+        [JsonProperty(PropertyName = "height")]
+        public int Height { get; set; }
+
+        /// <summary>
+        /// Height at start of activation window.
+        /// </summary>
+        [JsonProperty(PropertyName = "periodStartsHeight")]
```
Comment (quantumagi) [inline src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs]: `periodStartHeight`

File: src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+        public int PeriodStartsHeight { get; set; }
+
+        /// <summary>
+        /// Height at end of activation window.
+        /// </summary>
+        [JsonProperty(PropertyName = "periodEndsHeight")]
```
Comment (quantumagi) [inline src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs]: `periodEndHeight`

File: src/Stratis.Bitcoin/Base/Deployments/ThresholdConditionCache.cs [hunk 22]
```diff
@@ -58,6 +59,55 @@ public ThresholdState[] GetStates(ChainedHeader pindexPrev)
...
+
+            for (int deploymentIndex = 0; deploymentIndex < array.Length; deploymentIndex++)
+            {
+                if (this.consensus.BIP9Deployments[deploymentIndex] == null) continue;
+
+                DateTime? timeStart = this.consensus.BIP9Deployments[deploymentIndex]?.StartTime.Date;
+                DateTime? timeTimeout = this.consensus.BIP9Deployments[deploymentIndex]?.Timeout.Date;
+                int threshold = this.consensus.RuleChangeActivationThreshold;
+
+                int votes = 0;
```
Comment (quantumagi) [inline src/Stratis.Bitcoin/Base/Deployments/ThresholdConditionCache.cs]: The votes calculated here are the votes counted in the most recent window. If the state is already `Active` then this vote count may be irrelevant and **would not necessarily equal the number of votes that led to activation**.

File: src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+        public int PeriodEndsHeight { get; set; }
+
+        /// <summary>
+        /// Activation votes for this BIP9 deployment.
```
Comment (quantumagi) [inline src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs]: `Number of blocks with flags set for this BIP9 deployment in the last confirmation window.`

**Comment by dangershony** [comment]
Brilliant, I am ready to push this.
@noescape00 @bokobza please give this attention again so we can push to master.

File: src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs [hunk 24]
```diff
@@ -45,6 +49,35 @@ public uint256 GetBestBlockHashRPC()
...
+        /// <returns>A <see cref="JsonResult"/> object derived from a list of
+        /// <see cref="ThresholdStateModel"/> objects - one per deployment.
+        /// Returns an <see cref="ErrorResult"/> if the method fails.</returns>
+        [Route("api/[controller]/deploymentflags")]
+        [HttpGet]
+        public IActionResult DeploymentFlags()
+        {
+            try
+            {
+                ConsensusRuleEngine ruleEngine = this.ConsensusManager.ConsensusRules.GetRule<SetActivationDeploymentsFullValidationRule>().Parent;
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs]: this is a very inefficient way to get `ConsensusRuleEngine`. 
I think it's on DI so we can get it in the constructor of the controller

File: src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs (hunk 10 above)
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs]: Nope, it's not there

File: src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs (hunk 23 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs]: Done

File: src/Stratis.Bitcoin/Base/Deployments/ThresholdConditionCache.cs (hunk 22 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/Base/Deployments/ThresholdConditionCache.cs]: Done

File: src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs (hunk 21 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs]: Done

File: src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs (hunk 20 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs]: Done

File: src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs (hunk 19 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/Base/Deployments/Models/ThresholdStateModel.cs]: Done

File: src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs (hunk 18 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs]: Done

File: src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs [hunk 25]
```diff
@@ -101,6 +137,7 @@ public IActionResult GetBlockHashAPI([FromQuery] int height)
             }
             catch (Exception e)
             {
+                this.logger.LogTrace("(-)[EXCEPTION]");
```
[the remark this answers is not in the document]
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs]: Done

File: src/Stratis.Bitcoin/Base/Deployments/ThresholdConditionCache.cs (hunk 17 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/Base/Deployments/ThresholdConditionCache.cs]: Done

File: src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs (hunk 16 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs]: Done

File: src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs (hunk 15 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs]: Done

File: src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs (hunk 14 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs]: Done

File: src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs (hunk 13 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs]: Done

File: src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs (hunk 12 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs]: Done

File: src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs (hunk 24 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Consensus/ConsensusController.cs]: Done

**Comment by dangershony** [comment]
@quantumagi please merge master so tests can pass.

**Pull request merged**